fix brotli compression

sullis opened the pull request:
Motivation:

Brotli compressed http responses are not properly encoded.

Modification:

- change BrotliEncoder behavior so that the "finish" operation is performed before the channel is closed
- add new Brotli unit tests

Result:

All brotli tests pass.

sullis commented:
when I run HttpCompressTest locally, the Brotli test fails:

```
[ERROR] Failures: 
[ERROR]   HttpCompressionTest.brotliOnly:74 expected: <DONE> but was: <NEEDS_MORE_INPUT>
```

hyperxpro commented:
IMO, We don't really need a dedicated test suite for testing compression algos.

sullis commented:
I have a sample application that uses brotli4j **Encoder.compress** to compress the string:

```
byte[] compressed = Encoder.compress(inputText.getBytes(), StandardCompressionOptions.brotli().parameters());
System.out.println("inputText: " + inputText);
System.out.println("compressed length: " + compressed.length);
System.out.println("compressed: " + Arrays.toString(compressed));
```

The output is:

```
inputText: Hello Hello Hello Hello Hello
compressed length: 18
compressed: [27, 28, 0, 0, 4, 38, 114, -92, 25, 70, 114, 107, 123, -48, 19, -121, -72, 58]
```

When I run the HttpCompressionTest, the HTTP response body is different:

```
byte[] compressedData = IOUtils.toByteArray(inputStream);
System.out.println("inputText: " + TestConstants.CONTENT);
System.out.println("compressedData.length: " + compressedData.length);
System.out.println("compressedData: " + Arrays.toString(compressedData));
```

Console:

```
inputText: Hello Hello Hello Hello Hello
compressedData.length: 18
compressedData: [11, 14, 0, 0, 4, 38, 114, -92, 25, 70, 114, 107, 123, -48, 19, -121, -72, 58]
```

The unit test fails:

```
        DirectDecompress directDecompress = DirectDecompress.decompress(compressedData);
        assertEquals(DecoderJNI.Status.DONE, directDecompress.getResultStatus());
```

Result:

```
org.opentest4j.AssertionFailedError: 
Expected :DONE
Actual   :NEEDS_MORE_INPUT
```

sullis commented:
in my local environment, I added extra println statements to BrotliEncoder.java

BrotliEncoder finish method:
```
        System.out.println("BrotliEncoder.finish: channel.isOpen=" + ctx.channel().isOpen());
        System.out.println("BrotliEncoder.finish: channel.isActive=" + ctx.channel().isActive());
```

The console output is:
```
BrotliEncoder.finish: channel.isOpen=false
BrotliEncoder.finish: channel.isActive=false
```

BrotliEncoder Writer close method:
```
            System.out.println("BrotliEncoder Writer: channel.isOpen=" + this.ctx.channel().isOpen());
            System.out.println("BrotliEncoder Writer: channel.isActive=" + this.ctx.channel().isActive());
```

The console output is:
```
BrotliEncoder Writer: channel.isOpen=false
BrotliEncoder Writer: channel.isActive=false
```

As far as I can tell, the channel is closed before BrotliEncoder.Writer invokes brotliEncoderChannel.close()

This line of code:
```
    encode(EncoderJNI.Operation.FINISH);
```

executes after the channel is closed.

hyperxpro commented:
I get it now. Can you try to implement this close implementation in local env and see how things work?

https://github.com/netty/netty/blob/5cf0a8db102733384592357b3f86442f6ac11bd3/codec/src/main/java/io/netty/handler/codec/compression/JZlibEncoder.java#L323

sullis commented:
> I get it now. Can you try to implement this close implementation in local env and see how things work?

Thanks!  I just pushed a commit where I changed BrotliEncoder.java

The brotli test is passing now.  

I left a "TODO" comment in the code. I'm not sure if we should have a "boolean finished" field like the one found in JdkZlibEncoder.java

hyperxpro commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliEncoder.java:
```diff
@@ -171,7 +167,17 @@ public boolean isSharable() {
      * @throws IOException If an error occurred during closure
      */
-    public void finish(ChannelHandlerContext ctx) throws IOException {
+    private ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) throws IOException {
```
```suggestion
    private ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) throws IOException {
    public void finish(ChannelHandlerContext ctx) throws IOException {
```

sullis replied:
Resolved.

hyperxpro reviewed (changes requested):
Revert method signature. `finish(ChannelHandlerContext)` will create new promise by itself. Add another method called `finishEncode` as you did this time and it will accept `ChannelHandlerContext` and `ChannelPromise`.

So basically, `finish` will call `finishEncode`. This will work exactly the same and we will not break API.

sullis commented:
in my local environment, HttpCompressionTest is passing:

<img width="454" alt="image" src="https://user-images.githubusercontent.com/30938/215701521-1cd07def-432c-4fcf-b5c5-9a877ee25933.png">

hyperxpro commented:
We should remove this module entirely and implement extensive test cases right into `the codec` test package. You might wanna squash all commits of this PR to make it look clean.

sullis commented:
I just squashed the commits.   I need to sign off now.  I'll take another look at this PR tomorrow.

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliEncoder.java:
```diff
@@ -172,6 +168,20 @@ public boolean isSharable() {
+
+        finished = true;
+
+        END TODO */
```
remove ?

sullis replied:
Done

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliEncoder.java:
```diff
@@ -184,6 +194,28 @@ public void finish(ChannelHandlerContext ctx) throws IOException {
+            ctx.executor().schedule(new Runnable() {
+                @Override
+                public void run() {
+                    ctx.close(promise);
```
```suggestion
                    if (!promise.isDone()) {
                        ctx.close(promise);
                    }
```

normanmaurer commented on codec/src/main/java/io/netty/handler/codec/compression/BrotliEncoder.java:
```diff
@@ -184,6 +194,28 @@ public void finish(ChannelHandlerContext ctx) throws IOException {
+
+        if (!f.isDone()) {
+            // Ensure the channel is closed even if the write operation completes in time.
+            ctx.executor().schedule(new Runnable() {
```
store the returned `Future` and cancel when handle is removed etc ?

sullis replied:
I just added a 'handlerRemoved' method.  Please take a look and let me know if I'm on the right track

sullis commented:
> IMO, We don't really need a dedicated test suite for testing compression algos.

I removed the test suite.  I added new Brotli tests in **HttpContentCompressorTest**

sullis commented:
@normanmaurer @hyperxpro can we include this fix in the upcoming Netty release (4.1.88) ?

sullis commented:
Just squashed my commits.

sullis commented:
> Build failed:
> 
> ```
> linux-x86_64-java11-graal build
> ```
> 
> Error:
> 
> <img alt="image" width="521" src="https://user-images.githubusercontent.com/30938/216142365-6b94d965-9dc8-4a0b-b597-1fd36e8696cb.png">

Leak appears to be happening here:

```
HttpContentCompressorTest.testBrotliFullHttpResponse
```

Leak message:
```
LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
[53516](https://github.com/netty/netty/actions/runs/4066846362/jobs/7003576233#step:6:53517)
Recent access records: 
[53517](https://github.com/netty/netty/actions/runs/4066846362/jobs/7003576233#step:6:53518)
#1:
[53518](https://github.com/netty/netty/actions/runs/4066846362/jobs/7003576233#step:6:53519)
	io.netty.buffer.AdvancedLeakAwareByteBuf.order(AdvancedLeakAwareByteBuf.java:71)
[53519](https://github.com/netty/netty/actions/runs/4066846362/jobs/7003576233#step:6:53520)
	io.netty.buffer.CompositeByteBuf.newComponent(CompositeByteBuf.java:346)
[53520](https://github.com/netty/netty/actions/runs/4066846362/jobs/7003576233#step:6:53521)
	io.netty.buffer.CompositeByteBuf.addComponent0(CompositeByteBuf.java:287)
[53521](https://github.com/netty/netty/actions/runs/4066846362/jobs/7003576233#step:6:53522)
	io.netty.buffer.CompositeByteBuf.addComponent(CompositeByteBuf.java:265)
[53522](https://github.com/netty/netty/actions/runs/4066846362/jobs/7003576233#step:6:53523)
	io.netty.buffer.CompositeByteBuf.addComponent(CompositeByteBuf.java:222)
[53523](https://github.com/netty/netty/actions/runs/4066846362/jobs/7003576233#step:6:53524)
	io.netty.handler.codec.http.HttpContentCompressorTest.testBrotliFullHttpResponse(HttpContentCompressorTest.java:807)
```

sullis commented:
> Leak appears to be happening here:
> 
> ```
> HttpContentCompressorTest.testBrotliFullHttpResponse
> ```

I made a fix in **testBrotliFullHttpResponse** and **testBrotliChunkedContent**

```
        contentBuf.release();
```

sullis commented:
Resource leak is fixed.  CI build is green  ✅

hyperxpro reviewed (approved):
Thanks for the hard-work and debugging. :)

normanmaurer commented:
@sullis just rebased and pushed some change for the timeout handling. Once CI passes I will merge this one. Thanks a lot for all the work on this

normanmaurer commented:
@sullis thanks a lot ... merged

sullis commented:
Netty 4.1.88 SNAPSHOT is looking good:
https://github.com/Netflix/zuul/pull/1452
